### ISSUE TITLE
[overlays] [proposal] improvements 

### ISSUE THREAD

**Issue by tlouisse** [ing-bank/lion#36]
While building the datepicker, I encountered a few missing features/shortcomings in the Overlay System:

**- auto a11y for invokers**
The overlay system takes care of all accessibility, the only thing that it doesn't take care of is the attrs for global overlays (aria-haspop="dialog" and aria-expanded="false/true" (should be toggled programatically when global overlay is non-modal)

**- hidesOnOutsideClick support for global overlays.**
A modal overlay shouldn't be closeable, but a global overlay is not neccessarily a modal overlay and should thus support this feature like localOverlay does (example: material design datepicker).

**- No horizontal centering of dialog**
Currently, the global dialog opens to the left of the screen. I think the default should be horizontally centered?

**- No vertical centering of dialog**
Currently, it  opens ‘below the end of the page’. That means there is no relation to invoker element. Think the default should be either a fixed number from top of window or always vertically centered

**- Not closing on esc when clicking on backdrop -> focus on body**
A modal dialog allows to click on backdrop and then body becomes the `document.activeElement`.

**- elementToFocusAfterHide -> doesn’t work**
Didn't check thoroughly, but I think it's not working (not when provided as arg to show method, also not when provided as prop 'elementToFocusAfterHide' to config.)

**- Smarter defaults: always need to add these to my own styling:**
    - Position: relative so that it is positioned above backdrop
    - Background: white (Just like HTML body. You can override to transparent or other color if needed.)

**- Shadow dom encapsulation by default**
The only way to make content put in global overlay not leak styles, is to put it in the web component you created yourself.

Imagine this: 
```js
    const myCtrl = overlays.add(
      new ModalDialogController({
        contentTemplate: () => html`<div class="overlay"><div class="overlay-heading">bla</div></div>`,
      }),
    );
```
Now this overlay should be aware of its context. If it should be used in an app that could be placed in whatever legacy context, it can easily conflict when generic names are used. (a legacy app with `.overlay { background: yellow }` would cause inward leaking. 
Something like this already happened to one of our consumers. 

The current solution, always providing our own web component, requires a lot more boilerplate (it's a bit cumbersome if you want to provide an api for Subclassers in the input-datepicker component). Also, if you provide content projection inside your web component, you would be exposed to global scope again. I think providing one layer of encapsulation above this, would solve this problem.

Also, `.global-overlays` can be encapsulated by a shadow root to make sure it will never conflict with anything defined outside of it (this is actually the real conflict we encountered, when it was called 'overlay' and clashed with a 3rd party lib).

I think shadow dom encapsulation should be the default with an opt-out (if you want to hook in to a globally defined style lib, only for compatibility with legacy apps).

**- Styling not configurable**
I want to be able to change the colour, opacity, animation of the backdrop. Also want to override vertical placement etc. Making the global-overlays a wc, would make this easier and in line with other 'style patchings' for Subclassers.

@LarsDenBakker, @bashmish, @jorenbroekema, @daKmoR:
Could you please give your thoughts about the above?
Some of them might be design considerations and I might miss something in my proposed solutions?

**Comment by tlouisse** [comment]
@jorenbroekema : relevant to fix for input-datepicker

Must haves:
- No horizontal and vertical centering of dialog
- Not closing on esc when clicking on backdrop -> focus on body
- elementToFocusAfterHide -> doesn’t work

Currently solved within datepicker, but tech debt we should get rid off:
- auto a11y for invokers
- smarter defaults

**Comment by daKmoR** [comment]
#### Bugs to fix:
- Not closing on esc when clicking on backdrop -> focus on body
- auto a11y for invokers
- elementToFocusAfterHide -> doesn’t work

#### nice to have Features:
- hidesOnOutsideClick support for global overlays.
- Styling not configurable


#### imho not needed:
- No horizontal centering of dialog
- No vertical centering of dialog
=> should be done by the dialog... the overlay only provides a full page "overlay" what and where you are placed is the "dialogs" concern 
- Smarter defaults: always need to add these to my own styling:
  - Position: relative so that it is positioned above backdrop
  - Background: white (Just like HTML body. You can override to transparent or other color if needed.)
=> if you don't want to create it yourself you might want to extend a Base Dialog?

#### Discuss separately:
- Shadow dom encapsulation by default
=> let's have a completely separate discussion about it... I think I would need to see a POC as I currently don't understand how this could work? are we talking about auto generating shadow dom wrappers? or are we talking about the globally added entry point for all overlays?

**Comment by bashmish** [comment]
> #### imho not needed:
> * No horizontal centering of dialog
> * No vertical centering of dialog
>   => should be done by the dialog... the overlay only provides a full page "overlay" what and where you are placed is the "dialogs" concern

On one hand I agree with this and remember how we ended up with such approach, but I also ran into an issue with positioning my overlay frame components, unfortunately don't remember what exactly I could not achieve. I also don't like the fact that the backdrop visuals are hardcoded in the overlay system, they should be configurable and fully controlled by anyone who creates different types of overlays and frame components for them. I believe the positioning issue I had was related to default styles of the backdrop element. Both my issues and Thijs's issue should be tackled together.

**Comment by tlouisse** [comment]
@daKmoR:

> #### imho not needed:
> * No horizontal centering of dialog
> * No vertical centering of dialog
> * Smarter defaults: always need to add these to my own styling:

Regardless of whether we want to create web components, we should always provide some (functional) styling as a default. (it currently already does this with a opaque backdrop etc.) I personally regard all of the above 'functional' and ready for use in 90% of cases.
Defaults would be inspired by the platform:
https://developer.mozilla.org/en-US/docs/Web/HTML/Element/dialog


> #### Discuss separately:
> * Shadow dom encapsulation by default
>   => let's have a completely separate discussion about it... I think I would need to see a POC as I currently don't understand how this could work? are we talking about auto generating shadow dom wrappers? or are we talking about the globally added entry point for all overlays?

Good idea to have separate discussion for this. Could make an issue if you want. The global one should definitely have one imo. Then the content wrapper wouldn't need one, unless you need to make use of global styles somehow (in that case the global entry point should use content projection and another (optional)encapsulating shadow layer within).


@bashmish:

> both are tested and work as expected (I also double checked that in the storybook to see how it works in action):
the [esc] works actually. The real problem is you give the body focus by clicking the backdrop, and then the esc key is not captured. Focusing something outside modal dialog content should just not be possible (so the esc key handling is not the problem)

**Comment by jorenbroekema** [comment]
Personally I think it's a good idea to have defaults that make sense, but are overridable. 

This applies imo especially to the horizontal and vertical centering of the dialog frame because I imagine that for 90% of the cases people expect center-center position. I certainly expected it in our dialog to be honest.

Question is where to put these defaults.
1) In GlobalOverlayController, so when people don't want it, they will ALWAYS have to override. We can always export a even more naked variant as a separate controller if people really need it. RawGlobalOverlayController or something, dunno, just throwing out an idea.
2) One level lower, in a DefaultDialogController or LionDialogController or something, where we put our smart defaults for positioning, background, etc.

For now I'll start on auto a11y for invokers since that seems to be the thing we can all agree on is something that needs to happen. I will then see if I can confirm if/why Thijs experiences buggy behavior with hidesOnOutsideClick and elementToFocusAfterHide ;)

**Comment by bashmish** [comment]
> the [esc] works actually. The real problem is you give the body focus by clicking the backdrop, and then the esc key is not captured. Focusing something outside modal dialog content should just not be possible (so the esc key handling is not the problem)

I think there is too much stuff in this issue, maybe split it up. For this particular bug I would like to see a test failing because I still don't fully get what the bug is.

**Comment by bashmish** [comment]
@jorenbroekema we can add a possibility for the global overlay system to position things, but the default should be "no particular position", because default only makes sense when you know it will be like 90% of case or more, but in our situation we have multiple types of overlays which are build up on the overlays system, and they all have a different position so they are all equally important.

**Comment by jorenbroekema** [comment]
I mean, I would argue in >90% of cases it's center center apart from toast or when people want a completely specific custom position.

This is what it looks like in my head in terms of use cases
![image](https://user-images.githubusercontent.com/36734656/57631025-feb3f580-759e-11e9-9aaa-d059c9f34b14.png)

Am I missing anything here?

**Comment by bashmish** [comment]
people will not use GlobalOverlayController directly, it will only be used in the implementation of other types of overlays

Let's say there will be main use casesL dialogs and toasts and 2 places in the code where we need to set the position of the GlobalOverlayController: one needs centered (DialogOverlayController), the other one needs top-right (ToastOverlayController). So 50%/50%.

Then people will create like 100 different dialogs in their apps and a system of notifications based on just one toast, so it will be 99% agains 1%, but the position will be predefined for them already in the controllers, so this proportion does not matter for the underlying layer defaults. You can think of it like if the default is set in the controller extension rather than in the base class.

**Comment by tlouisse** [comment]
I like the idea of a totally unbiased super class.
But then we should probably also remove the opaque black backdrop and its animation? What's the current recommended approach of configuring the '. global-overlay'? Overriding styles on body level?

**Comment by bashmish** [comment]
@tlouisse I agree with the idea of changing the backdrop implementation like I said above. The current way it is done is not flexible at all. I think this is one of the places where we need new `:part` functionality.

**Comment by tlouisse** [comment]
Created an overview of all possible Overlay occurrences to make.
It also addresses the positioning of the (modal) dialog 
Goal is to make it more clear what is the total end scope that should be handled by our generic layers: https://github.com/ing-bank/lion/blob/feat/richSelect/packages/overlays/docs/Occurrences.md

Feedback welcome ;)

**Comment by tlouisse** [comment]
> #### Discuss separately:
> * Shadow dom encapsulation by default
>   => let's have a completely separate discussion about it... I think I would need to see a POC as I currently don't understand how this could work? are we talking about auto generating shadow dom wrappers? or are we talking about the globally added entry point for all overlays?

Created a failing code example here: 
https://stackblitz.com/edit/global-overlays-in-and-outward-style-leaks?file=index.ts

About the generation of shadow dom, we could probably do it here, like:

```js
export class GlobalOverlayController {
  static _createRoot() {
    if (!this._rootNode) {
      this._rootNode = document.createElement('div');
      // TODO: create a shadow dom root here. Currently, the only way this works with shadyCSS is
      // by creating webcomponent, so create <lion-overlay-root> instead of <div>
      // Also, make this root configurable, so:
      // - it supports all viewport placements:
      //   {String} viewportPlacement - consists of 'v-align' (vertical alignment) and 'h-align' (horizontal alignment), separated via '-'
      //     - v-align : 'center' | 'bottom' | 'top' | 'left' | 'right' | 'fullheight'
      //     - h-align: 'middle' | 'start' | 'end' | 'fullwidth'
      // - allow to style backdrop (opacity, color, animation(?))
      // - allow to animate overlay
      this._rootNode.classList.add('global-overlays');
      document.body.appendChild(this._rootNode);
      // TODO: add part to web component described above, and parts to body
      document.head.appendChild(styleTag);
    }
  }
```

**Comment by erikkroes** [comment]
Hi,

We are closing this issue because we're changing the way we handle feature requests. Your issue will get the "votes needed" label, and people can add 👍 reactions to [vote and show interest](https://github.com/ing-bank/lion/issues?q=label%3A%22votes+needed%22+sort%3Areactions-%2B1-desc). It will be reopened when this feature gets picked up. 

This we way we clearly focus on bugs in our open issues. At the same time we can see what community interest there is for new features.

Thanks!

**Comment by daKmoR** [comment]
This should all be implemented? if something is missing then please open a dedicated issue for it.

going to "close" this feature request.
